parley/src/analysis/mod.rs

```diff
@@ -0,0 +1,2011 @@
+        if wb == text.len() {
+            continue;
+        }
+        all_boundaries_byte_indexed[wb] = Boundary::Word;
+    }
+
+    // Line boundaries (word break naming refers to the line boundary determination config).
+    //
+    // This breaks text into sequences with similar line boundary config (part of style
+    // information). If this config is consistent for all text, we use a fast path through this.
+    let Some((first_style, rest)) = lcx.styles.split_first() else {
+        panic!("No style info");
+    };
+    let contiguous_word_break_substrings = WordBreakSegmentIter::new(
+        text,
+        rest.iter(),
+        &first_style
+    );
+    let mut global_offset = 0;
+    for (substring_index, (substring, word_break_strength, last)) in contiguous_word_break_substrings.enumerate() {
+        let line_boundaries: Vec<usize> = lcx.analysis_data_sources
+            .line_segmenter(word_break_strength)
+            .segment_str(substring)
+            .collect();
+
+        // Fast path for text with a single word-break option.
+        if substring_index == 0 && last {
+            // icu adds leading and trailing line boundaries, which we don't use.
+            let Some((_first, rest)) = line_boundaries.split_first() else {
+                continue;
+            };
+            let Some((_last, middle)) = rest.split_last() else {
+                continue;
+            };
+            for &b in middle {
+                all_boundaries_byte_indexed[b] = Boundary::Line;
+            }
+            break;
+        }
+
+        let mut substring_chars = substring.chars();
+        if substring_index != 0 {
+            global_offset -= substring_chars.next().unwrap().len_utf8();
+        }
+        // There will always be at least two characters if we are not taking the fast path for
+        // a single word break style substring.
+        let last_len = substring_chars.next_back().unwrap().len_utf8();
+
+        // Mark line boundaries (overriding word boundaries where present).
+        for (index, &pos) in line_boundaries.iter().enumerate() {
+            // icu adds leading and trailing line boundaries, which we don't use.
+            if index == 0 || index == line_boundaries.len() - 1 {
+                continue;
+            }
+
+            // For all but the last substring, we ignore line boundaries caused by the last
+            // character, as this character is carried back from the next substring, and will be
+            // accounted for there.
+            if !last && pos == substring.len() - last_len {
+                continue;
+            }
+            all_boundaries_byte_indexed[pos + global_offset] = Boundary::Line;
+        }
+
+        if !last {
+            global_offset += substring.len() - last_len;
+        }
+    }
+
+    // BiDi embedding levels:
+    let bidi_embedding_levels = unicode_bidi::BidiInfo::new_with_data_source(&lcx.analysis_data_sources.bidi_class(), text, None).levels;
```

Comment:
This bidi impl doesn't enable passing in allocations (it allocates within its impl). I created an issue in the upstream repo asking whether it's possible to pass allocations in: https://github.com/servo/unicode-bidi/issues/146